processor/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,504 @@
+		dpCalls.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+		dpCalls.SetIntVal(c)
+
+		dimensions, ok := p.dimensionsForSeries(key)
+		if !ok {
+			return fmt.Errorf("failed to find dimensions for key %s", key)
+		}
+
+		dimensions.CopyTo(dpCalls.Attributes())
+	}
+
+	return nil
+}
+
+func (p *processor) collectLatencyMetrics(ilm pmetric.ScopeMetrics) error {
+	for key := range p.reqDurationSecondsCount {
+		mDuration := ilm.Metrics().AppendEmpty()
+		mDuration.SetDataType(pmetric.MetricDataTypeHistogram)
+		mDuration.SetName("request_duration_seconds")
+		// TODO: Support other aggregation temporalities
+		mDuration.Histogram().SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
+
+		timestamp := pcommon.NewTimestampFromTime(time.Now())
+
+		dpDuration := mDuration.Histogram().DataPoints().AppendEmpty()
+		dpDuration.SetStartTimestamp(pcommon.NewTimestampFromTime(p.startTime))
+		dpDuration.SetTimestamp(timestamp)
+		dpDuration.SetExplicitBounds(pcommon.NewImmutableFloat64Slice(p.reqDurationBounds))
+		dpDuration.SetBucketCounts(pcommon.NewImmutableUInt64Slice(p.reqDurationSecondsBucketCounts[key]))
+		dpDuration.SetCount(p.reqDurationSecondsCount[key])
+		dpDuration.SetSum(p.reqDurationSecondsSum[key])
+
+		// TODO: Support exemplars
+
+		dimensions, ok := p.dimensionsForSeries(key)
+		if !ok {
+			return fmt.Errorf("failed to find dimensions for key %s", key)
+		}
+
+		dimensions.CopyTo(dpDuration.Attributes())
+	}
+	return nil
+}
+
+func (p *processor) buildMetricKey(clientName, serverName, connectionType string, edgeDimensions map[string]string) string {
+	var metricKey strings.Builder
+	metricKey.WriteString(clientName + metricKeySeparator + serverName + metricKeySeparator + connectionType)
+
+	for _, dimName := range p.config.Dimensions {
+		dim, ok := edgeDimensions[dimName]
+		if !ok {
+			continue
+		}
+		metricKey.WriteString(metricKeySeparator + dim)
+	}
+
+	return metricKey.String()
+}
+
+// cacheLoop runs in a separate goroutine and periodically cleans the cache
+func (p *processor) cacheLoop(d time.Duration) {
+	t := time.NewTicker(d)
+	for {
+		select {
+		case <-t.C:
+			p.cleanCache()
+		case <-p.shutdownCh:
+			return
+		}
+	}
+
+}
+
+// cleanCache removes series that have not been updated in 15 minutes
+func (p *processor) cleanCache() {
+	var staleSeries []string
+	for key, series := range p.keyToMetric {
+		if series.lastUpdated+15*time.Minute.Milliseconds() < time.Now().UnixMilli() {
+			staleSeries = append(staleSeries, key)
+		}
+	}
+
+	for _, key := range staleSeries {
+		delete(p.keyToMetric, key)
+	}
+}
+
+func buildEdgeKey(k1, k2 string) string {
+	var b strings.Builder
+	b.WriteString(k1)
+	b.WriteString("-")
+	b.WriteString(k2)
+	return b.String()
+}
+
+// durationToMillis converts the given duration to the number of milliseconds it represents.
+// Note that this can return sub-millisecond (i.e. < 1ms) values as well.
+func durationToMillis(d time.Duration) float64 {
```

Comment:
Here it is! The readme says this would be seconds, but it looks like it is milliseconds. In any case, I think the config type can either be changed to be a slice of integers or the conversion should be removed altogether. Right now, it's a slice of durations, so users can in theory specify "1d" and this will be converted to float here.

Reply:
This should be fixed. The config admits a slice of `time.Duration`. Converting it to milliseconds makes it easier to work with metrics.